Marlin/src/HAL/STM32/fast_pwm.cpp

```diff
@@ -79,6 +79,12 @@ void set_pwm_frequency(const pin_t pin, int f_desired) {
 
   uint32_t index = get_timer_index(Instance);
 
+  // Protect used timers
+  if (index == TEMP_TIMER_NUM || index == STEP_TIMER_NUM) return;
```

Comment:
beware with that, get_timer_index returns 0 for TIM1, TEMP_TIMER_NUM & STEP_TIMER_NUM seems to put 1 for TIM1

Reply:
index = get_timer_index(Instance) + 1; may work, and put 0 for UNKNOWN_TIMER (only with uint16_t), not with your uint32_t, will be 0xFFFF + 1 = 0x10000

Reply:
Thanks for the heads up. 
Yes, I accounted for that, the defines are set to the index in timers.h
e.g. #define TEMP_TIMER_NUM 0 //TIM1
The entire library uses 0 as the first timer. 
This code is what was was used in the __STM32F1__ HAL
The library uses a 32 bit var
See https://github.com/stm32duino/Arduino_Core_STM32/blob/d217150fd64476cfb0d7a14f80f89d80918eada5/cores/arduino/HardwareTimer.cpp

Reply:
now im a bit confused about TEMP_TIMER_NUM & STEP_TIMER_NUM if you say those are not TIMER real nums, i dont believe so

Reply:
I will demonstrate with debug logging.

Reply:
k, #define TEMP_TIMER              3
// Leave TIMER 4 for TFT backlight PWM or Servo freq...
#define STEP_TIMER              5

no _NUM suffix here

Reply:
#define __TIMER_DEV(X) TIM##X
#define _TIMER_DEV(X) __TIMER_DEV(X)
#define STEP_TIMER_DEV _TIMER_DEV(STEP_TIMER)
#define TEMP_TIMER_DEV _TIMER_DEV(TEMP_TIMER)

_NUM ones are internal to marlin (only) offsets, so only 2.... 0 and 1

#ifndef STEP_TIMER_NUM
  #define STEP_TIMER_NUM        0  // Timer Index for Stepper
#endif
#ifndef PULSE_TIMER_NUM
  #define PULSE_TIMER_NUM       STEP_TIMER_NUM
#endif
#ifndef TEMP_TIMER_NUM
  #define TEMP_TIMER_NUM        1  // Timer Index for Temperature
#endif
its related to #define NUM_HARDWARE_TIMERS 2
HardwareTimer *timer_instance[NUM_HARDWARE_TIMERS] = { nullptr };

Reply:
timers.h of STM32
#ifndef STEP_TIMER_NUM
  #define STEP_TIMER_NUM        0  // Timer Index for Stepper
#endif
#ifndef PULSE_TIMER_NUM
  #define PULSE_TIMER_NUM       STEP_TIMER_NUM
#endif
#ifndef TEMP_TIMER_NUM
  #define TEMP_TIMER_NUM        1  // Timer Index for Temperature
#endif

Reply:
you cant compare get_timer_index which can be 0 to 21 with STEP_TIMER_NUM & TEMP_TIMER_NUM which can only be 0 & 1

Reply:
The function is to protect Marlin system timers not all timers.

Reply:
yep but the test is inccorect, you will only avoid the use of TIM1 & TIM2 and protect nothing

Reply:
I will test it with logging to verify, one moment.

Reply:
you can use this array:
static constexpr struct { TimerPurpose p; int t; } timers_in_use[] = {
  #if HAS_TMC_SW_SERIAL
    {TP_SERIAL, get_timer_num_from_base_address(timer_serial[0])},  // Set in variant.h, or as a define in platformio.h if not present in variant.h
  #endif
  #if ENABLED(SPEAKER)
    {TP_TONE, get_timer_num_from_base_address(timer_tone[0])},    // Set in variant.h, or as a define in platformio.h if not present in variant.h
  #endif
  #if HAS_SERVOS
    {TP_SERVO, get_timer_num_from_base_address(timer_servo[0])},   // Set in variant.h, or as a define in platformio.h if not present in variant.h
  #endif
  {TP_STEP, STEP_TIMER},
  {TP_TEMP, TEMP_TIMER},
};

Reply:
I understand, you would like to expand beyond the current system critical timers that were previously defined. However I think that's beyond the intent of this PR but yes we can look at that. The other HALs do not go beyond these as far as I can tell. For example LCP1768 has no protection.

Reply:
yes and no, just want a working test/code :p

Reply:
For sure, this is good.

Reply:
The test does pass:

in set_pwm_frequency we log 
 uint32_t index = get_timer_index(Instance);
  SERIAL_ECHO_MSG("GetTimerIndex",index);
  SERIAL_ECHO_MSG("TEMP_TIMER_NUM",TEMP_TIMER_NUM);
  SERIAL_ECHO_MSG("STEP_TIMER_NUM",STEP_TIMER_NUM);
  // Protect used timers
  if (index == TEMP_TIMER_NUM || index == STEP_TIMER_NUM) return;
  SERIAL_ECHO_MSG("IndexTestFailed",TEMP_TIMER_NUM);

we call set_pwm_frequency(pin_name, 2000);

this is the output

10:52:40.590 > echo:GetTimerIndex0
10:52:40.590 > echo:TEMP_TIMER_NUM1
10:52:40.598 > echo:STEP_TIMER_NUM0

Reply:
This code does the intended protection at the moment. 
We would need to define a global scope of what is considered critical in any expansion.
Like a registry.

Reply:
> these defines are not configurable, just offsets for the marlin custom timers (HT) storage
Based on the logs, the index does match up to the TIMx instance. It is not what you think it's doing. The HT pointer is a 32 bit handle and not related to index. 
e.g.
from the lib HardwareTimer.cpp

timer_index_t get_timer_index(TIM_TypeDef *instance)
{
  timer_index_t index = UNKNOWN_TIMER;

#if defined(TIM1_BASE)
  if (instance == TIM1) {
    index = TIMER1_INDEX;
  }
#endif 
...

Reply:
i just said TEMP_TIMER_NUM is 1 for all STM32 printers and STEP is always 0, and index can be up to 21, you dont compare the same "indexes"

Reply:
Ahh, I see.

Reply:
So yes, we can go beyond these 2 previously protected timers, the question is do we do that now and take more time on this PR or do we do another PR to look at all the HALs.

Reply:
or use a very different way : https://github.com/tpruvot/Marlin/commit/7170df229c7d3b0b187e451ecd126366b01773db

Reply:
I don't agree with the method shown for the following reasons:

1) Duplicates background Arduino managed calls into Marlin.
2) Entangles Marlin FAN function in the calls.
3) Calls digitalWrite when hardware is exclusively required for the call.
4) Does not account for other use cases.

I do agree with the method shown for the following reasons:

1) Enumeration of existing set timer frequencies into a registry array.